### DISCUSSION TITLE
Question with Filters - Integrating and Resampling Data and filterung x by datetime

### DISCUSSION THREAD

**Discussion by marce14git** [dbuezas/lovelace-plotly-graph-card#307] [Q&A] [answered]
Hey there, 

I am sorry to bother you again, because i am not able to resolve my problem by myself.
as discussed in the HA Forum 
[HA-Link](https://community.home-assistant.io/t/plotly-interactive-graph-card/347746/374?u=marce14-git)

I have 2 Problems:

1. i have a local precip sensor which gets roughly every 5min a value, but i would like to sum those values up and display them only every hour, like the values from openweather for better visual comparison

```
   filters:
      - integrate: 
           unit: h
           reset_every: 1h
      - resample: 1h
      - multiply: 12
```

As suggested i tried to use the integrate filter and resample but i get completly wrong values 
<img width="398" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/92689239/cddc0c5b-280b-402f-861f-9b48aeda1339">

in the screenshot below are the raw numbers if i delete the filters from above (no recent values before the displayed below)

<img width="391" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/92689239/50edb3ac-e8c0-4c5c-8830-9b8184774302">

2. Question:
as suggested i dont want to display values which are older than NOW 
<img width="366" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/92689239/73f61834-fcba-4ec6-95b0-d03c43f379db">

so i tried using filters

```
    filters:
      - store_var: precipitation_p
      - fn: |-
          ({ meta, vars }) => ({
            xs: [vars.precipitation_p.xs.slice(-1)[0],...meta.forecast.map(({ datetime }) => new Date(datetime))],
            ys: [vars.precipitation_p.ys.slice(-1)[0],...meta.forecast.map(({ precipitation_probability }) => precipitation_probability)],
          })
      - filter: x > Date.now
```

BUT if i use this filter the whole line DISAPPEARS
<img width="390" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/92689239/25777968-ea21-4ff2-845a-e20e0a60bed3">

Question 2 is only a minor annoyance but on paper the fix seems easy, if it would work 

thanks very much in advance

**Comment by dbuezas** [comment]
Hello there!

### For problem 1

Let's go through it: if the report is every 5 minutes, I assume that you get the total mm rained in the last 5 minutes only.
When you integrate, the system assumes the value is returning "rate of rain per hour", whereas the data actually has the "rate of rain per 5 minutes" (or total rain in the last 5 minutes).
Therefore, multiplying by 12 (60/5) should give you the correct results. Please try duplicating the trace, but this time removing the `resample` and check if the last values of each hour look correct and post some screenshots. If they do, then the resample may be picking the value on the wrong place.  

### For problem 2

`Date.now` is actually a function, so you need to call it:
`filter: x > Date.now`  ==> `filter: x > Date.now()`

---

**Reply by marce14git** [reply]
thanks for your swift answer

ad 2:
Thanks - i am a moron :D it works now, THANKS!


ad 1:
```
    filters:
      - integrate: 
           unit: h
           reset_every: 1h      
      - resample: 1h
      - multiply: 12
```

<img width="384" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/92689239/52487b6a-34b9-4014-9076-f39999140d67">


<img width="371" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/92689239/7f301b58-2832-4494-b937-75c3fde9baca">

so the RED bar at 11 is correct its 4 (there is 1 value=1 behind an orange bar at 10:35)
BUT all the other small red bars before are false! there was 0,1mm rain at 4:40 and then nothing until 10:10 where i inserted some values for testing


thanks in advance

**Reply by dbuezas** [reply, marked as answer]
Oh I think I know what's going on. Can you plot it as lines instead of bars? this hides unavailable datapoints.
Try `resample: 5m` BEFORE integrating to ensure there are datapoints everywhere.

**Reply by dbuezas** [reply]
(I mean adding the 5 min resample before integrating, and leaving the 1h one after)

**Comment by marce14git** [comment]
Oh wow, i didnt even know u could use multiple resamples

i checked, and i get datapoints every 10min +- a few seconds (i guess thats the reason its somtimes a a tiny bit off), but nevertheless you solved the problem  <3 

thank you so much

<img width="367" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/92689239/c2bda97c-b84d-4e1f-a9c0-f2de9c0d314a">

**Reply by dbuezas** [reply]
Sure you can! The filters are a pipeline, they are applied in order, each filter is a function that receives the output of the previous one. You can even run code inside them (an also outside, but let's go step by step :)

**Reply by marce14git** [reply]
again thank you so much and sorry for asking again, i searched all topics on github and the HA forum but i cant get it done...

I am trying to replicate my Apexcharts graph(was slowing my HA on mobile down) 
<img width="386" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/92689239/534b21ed-808c-4adb-86ec-f3b8e7487d9a">

the last thing that is missing is the marker/annotation/label of the min and max temperature - is that possible? and if yes would it be possible to show the min/max of the last 24h at the line graph?

**Reply by dbuezas** [reply]
Hi Marce,
there's no pre made feature for this, but it can be achieved with an extra trace and filters. I'll give it a try and get back here when I have time.

**Reply by dbuezas** [reply]
Can you put this in a new Q&A discussion please? It will be easier for others to find it by title

**Reply by marce14git** [reply]
of course and again thank you very much [new Q&A](https://github.com/dbuezas/lovelace-plotly-graph-card/discussions/309)

**Reply by dbuezas** [reply]
Thanks! I'll take a look soon :)
